AK/SyncGenerator.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2025, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Coroutine.h>
+#include <AK/Optional.h>
+#include <AK/ScopeGuard.h>
+
+namespace AK {
+
+template<typename Y>
+class SyncGenerator {
+    struct PromiseType;
+
+    AK_MAKE_NONCOPYABLE(SyncGenerator);
+
+public:
+    using YieldType = Y;
+    using promise_type = PromiseType;
+
+    ~SyncGenerator()
+    {
+        if (!m_handle)
+            return;
+
+#ifdef AK_COROUTINE_DESTRUCTION_BROKEN
+        // FIXME: Calling `m_handle.destroy()` when a coroutine has not reached its final
+        //        suspension point is scary on Clang < 19.
+        while (!m_handle.done()) {
+            m_handle.resume();
+            m_value.~Y();
+        }
+#endif
+
+        m_handle.destroy();
+    }
+
+    SyncGenerator(SyncGenerator&& other)
+        : m_handle(AK::exchange(other.m_handle, {}))
+    {
+        if (m_handle)
+            m_handle.promise().m_generator = this;
+    }
+
+    SyncGenerator& operator=(SyncGenerator&& other)
+    {
+        if (this != &other) {
+            this->~SyncGenerator();
+            new (this) SyncGenerator(move(other));
+        }
+        return *this;
+    }
+
+    bool is_done() const
+    {
+        return m_handle.done();
+    }
+
+    Optional<Y> next()
```

Comment:
the normal design here is to `VERIFY(!is_done())` (i.e. require that this is only called if `!is_done()` and return a non-optional, right? Could we do that?